python/utils.py

```diff
@@ -55,8 +76,95 @@ def read_data_file(path="", data_file=None):
+				columns=lar_schema.field)
 			return (ts_df, lar_df)
 	else:
-		raise ValueError("A data file mus be passed. No data have been written to the object")
\ No newline at end of file
+		raise ValueError("A data file mus be passed. No data have been written to the object")
+
+def unique_uli(new_lar_df=None, lei=None):
+    """
+    Generates a set of unique ULI's for a LAR dataframe.
+    """
+
+    #Copying over ULIs with the LEI.
+    new_lar_df["uli"] = new_lar_df["uli"].apply(lambda x: 
+        lei)  
+    
+    #Generates a loan ID as a random 23-character 
+    #string for each LEI.
+    new_lar_df["uli"] = new_lar_df["uli"].apply(lambda x: x + 
+        lar_gen.char_string_gen(23)) 
+    
+    #Adds a check digit to each.
+    new_lar_df['uli'] = new_lar_df["uli"].apply(lambda x: x + 
+        lar_gen.check_digit_gen(ULI=x))
+
+    #If ULIs are duplicated, the unique_uli function 
+    #is applied again.  
+    if len(new_lar_df['uli']) > len(set(new_lar_df['uli'])):
+        print("Re-Running")
+        self.unique_uli(new_lar_df)
+    else:
+        print("Unique ULIs Assigned")
+
+    return new_lar_df
+
+def new_lar_rows(row_count=None, lar_df=None, ts_df=None):
+    """
+    Returns LAR with a specified number of rows. 
+    Returns TS data with updated lar entries.
+    """
+
+    #Stores the LAR dataframe as a new LAR dataframe. 
+    new_lar_df = lar_df
+
+    #Stores the number of rows currently in the dataframe to calculate the 
+    #new number of rows to concatenate. 
+    current_row = len(new_lar_df.index) 
+    
+    #Calculates a multiplier taking the ceiling function of 
+    #the desired row count over the current row count.
+    multiplier = math.ceil(row_count/current_row)
+    
+    #Concatenates data to produce the number of rows 
+    #by the multiplier in a new LAR dataframe. 
+    new_lar_df = pd.concat([new_lar_df]*int(multiplier))
+    
+    #Drops the number of rows to the count specified. 
+    if (row_count % current_row) != 0:
+        drop_rows = current_row - (row_count % current_row)
+        new_lar_df = new_lar_df[: - (drop_rows)]
+    
+    #Applies the unique_uli function to the new LAR dataframe 
+    #to generate a unique set of ULIs. 
+    new_lar_df = utils.unique_uli(new_lar_df=new_lar_df, 
+    	lei=ts_df.iloc[0][14])
+
+    #Modifies TS data for the new number of LAR entries.
+    ts_df["lar_entries"] = len(new_lar_df)
+
+    return (new_lar_df, ts_df)
+
+def row_by_row_modification(lar_df=None):
+	"""Uses the inputs from the row_by_row 
+	modification yaml to modify a dataframe"""
+
+	#Opens the yaml_file. 
+	yaml_file ='row_by_row_modification.yaml'
+	with open(yaml_file, 'r') as f:
+		row_by_row = yaml.load(f)
+	
+	#Runs through each case in the file. 
+	#Modifies values for each column in a row specified in the yaml file. 
+	for case in row_by_row:
+		for column in row_by_row[case]["columns"]:
+			for key in column:
+				lar_df.iloc[row_by_row[case]["row"]][key] = column[key]
```

Comment:
Check out the df.at function. It's a bit cleaner than chained .iloc calls. Helps avoid chained index issues.